client/components/infinite-scroll/index.js

```diff
@@ -3,53 +3,92 @@
 /**
  * External dependencies
  */
-import { Component } from 'react';
+import React from 'react';
 import PropTypes from 'prop-types';
-import { throttle } from 'lodash';
+import { throttle, defer } from 'lodash';
+import afterLayoutFlush from 'lib/after-layout-flush';
 
 const SCROLL_CHECK_RATE_IN_MS = 400;
 
-export default class InfiniteScroll extends Component {
+export default class InfiniteScroll extends React.Component {
 	static propTypes = {
 		nextPageMethod: PropTypes.func.isRequired,
 	};
 
-	checkScrollPositionHandler = throttle(
-		() => this.checkScrollPosition( true ),
-		SCROLL_CHECK_RATE_IN_MS
-	);
+	observedElement = React.createRef();
 
 	componentDidMount() {
-		window.addEventListener( 'scroll', this.checkScrollPositionHandler );
-		this.checkScrollPosition( false );
+		this.hasIntersectionObserver = 'IntersectionObserver' in window;
+
+		if ( this.hasIntersectionObserver && this.observedElement.current ) {
+			this.observer = new IntersectionObserver( this.handleIntersection, {
+				rootMargin: '100%',
+				threshold: 1.0,
+			} );
+			this.observer.observe( this.observedElement.current );
+		} else {
+			window.addEventListener( 'scroll', this.checkScrollPositionHandler );
+			this.deferredTimer = defer( () => this.checkScrollPositionHandler( false ) );
+		}
 	}
 
 	componentDidUpdate() {
-		this.checkScrollPosition( false );
+		if ( ! this.hasIntersectionObserver && ! this.deferredTimer ) {
+			this.deferredTimer = defer( () => this.checkScrollPositionHandler( false ) );
+		}
 	}
 
 	componentWillUnmount() {
+		// Clean up scroll codepath.
 		window.removeEventListener( 'scroll', this.checkScrollPositionHandler );
+		this.clearTimers();
```

Comment:
After `this.clearTimers()`, there can still be dangling scheduled callbacks after unmounting. They are scheduled by `throttle` and `afterLayoutFlush` in `checkScrollPositionHandler`. Both wrappers return a function decorated with a `cancel` method that can be used to cancel them.

Reply:
Very good point, we want to avoid that. Should be fixed now.